### DISCUSSION TITLE
Quay authentication issue

### DISCUSSION THREAD

**Discussion by giokara-work** [renovatebot/renovate#10188] [Request Help]
Recently we started seeing authentication issues for quay.io private repositories. Excerpt from the logs:

```
DEBUG: Datasource unauthorized
{
  "datasource": "docker",
  "lookupName": "quay.io/...",
  "url": "https://quay.io/api/v1/repository/.../tag/?limit=100&page=1&onlyActiveTags=true"
}
DEBUG: Failed to look up dependency quay.io/... (quay.io/...)(packageFile=".../Dockerfile", dependency="quay.io/...")
```

What could be possible causes of this? It seemed related to changes in https://github.com/renovatebot/renovate/pull/10093 which treats authentication for quay and other docker repositories differently. Therefore I opened https://github.com/renovatebot/renovate/pull/10185 as a possible fix. As pointed out by @viceice however, the authentication should be treated at http layer and therefore https://github.com/renovatebot/renovate/pull/10185 is not necessary.

**Comment by viceice** [comment]
Do you try to update a private dependency?

Please provide some more context log lines, the failed requests should be looged before those messages.

**Reply by giokara-work** [reply]
This is indeed for private docker dependencies. It has worked successfully in the past.

Some more logs that might be relevant:
```
DEBUG: Found encrypted config
{
  "config": {
    "password": "***********",
    "username": "..."
  }
}
DEBUG: Trying default padding for password
DEBUG: Trying RSA_PKCS1_PADDING for password
DEBUG: Decrypted password
DEBUG: Trying default padding for username
DEBUG: Trying RSA_PKCS1_PADDING for username
DEBUG: Decrypted username
DEBUG: Setting hostRules from config
DEBUG: Adding password authentication for https://quay.io to hostRules
```

The error in the description starts appearing directly after `INFO: Dependency extraction complete`. Could not find anything more of relevance. Happy to provide more info if you are looking for something specific.

**Reply by viceice** [reply]
OK, i found the difference: https://docs.quay.io/api/#making-an-api-request

for docker you need `username` and `apitoken` as `password` login, but for api requests you need a `apitoken` as `token` in hostrule.

So you need to configure this host rules:
```json
{
  "hostRules": [
    {
      "hostType": "docker",
      "matchHost": "https://quay.io/v2/",
      "username": "usr",
      "password": "apitoken"
    },
    {
      "hostType": "docker",
      "matchHost": "https://quay.io/api/v1/",
      "token": "apitoken"
    }
  ]
}
```

You can use `encryped` sub objects of cause. Those rules will cause the right auth assigning to the http requests.

@HonkingGoose Can you please add this sample to the docs if it is verified working.

**Reply by giokara-work** [reply]
Still the same with the following:
```
"hostRules": [
  {
    "hostType": "docker",
    "matchHost": "https://quay.io/v2/",
    "encrypted": {
      "password": "...",
      "username": "..."
    }
  },
  {
    "hostType": "docker",
    "matchHost": "https://quay.io/api/v1/",
    "encrypted": {
      "password": "...",
      "username": "..."
    }
  }
]
```
If I understand correctly I can just use the encrypted "password" from the v2 call as encrypted "token" of the v1 call? Will try this as well.

**Reply by viceice** [reply]
You need to use `token` instead of `username` and `password` as i told you above

```json
"hostRules": [
  {
    "hostType": "docker",
    "matchHost": "https://quay.io/v2/",
    "encrypted": {
      "password": "...",
      "username": "..."
    }
  },
  {
    "hostType": "docker",
    "matchHost": "https://quay.io/api/v1/",
    "encrypted": {
      "token": "..."
    }
  }
]
```

use the value from `password` for `token`

**Reply by giokara-work** [reply]
Still the same error is returned. Created a test repo to showcase the issue https://github.com/giokara/renovate-test and removed the credential encryption for now.

**Reply by giokara-work** [reply]
This also does not work: `curl -H "Authorization: Bearer <token>" https://quay.io/api/v1/repository/giokara/test/tag/\?limit\=100\&page\=1\&onlyActiveTags\=true`. Will look into some more detail in the quay authentication.

**Reply by viceice** [reply]
For api you need to use Bearer auth instead of basic. That's why I suggested to use token. 🤔

**Reply by giokara-work** [reply]
Indeed, adjusted in the above call (had tried both while testing). Still no success with tests through curl, have contacted Red Hat Quay support.

**Reply by viceice** [reply]
Maybe your token is missing some scopes? https://docs.quay.io/api/#generating-a-token-for-internal-application-use

You can't use your normal password as token for the quay api.

**Reply by giokara-work** [reply]
OK so I got already further.
- In a `hostRule` for `matchHost: https://quay.io/v2/` the user and password of a [quay robot account](https://docs.quay.io/glossary/robot-accounts.html) should be used.
- In a different `hostRule` for `matchHost: https://quay.io/api/v1/` the [OAuth token of an organisation application](https://docs.quay.io/api/#generating-a-token-for-internal-application-use) should be used as you mention.

**Reply by giokara-work** [reply]
So the request to `https://quay.io/api/v1/repository/giokara/test/tag/\?limit\=100\&page\=1\&onlyActiveTags\=true` now works correctly, but the manifests fetch now doesn't seem to use the quay.io/v2 credentials:
```
DEBUG: getManifestResponse(https://quay.io, <project>, latest)
DEBUG: Unauthorized docker lookup
{
  "registryHost": "https://quay.io",
  "dockerRepository": "<project>"
}
DEBUG: Response code 401 (UNAUTHORIZED)
{
  "err": {
    "name": "HTTPError",
    "timings": {
      "start": 1622209726861,
      "socket": 1622209726861,
      "lookup": 1622209726862,
      "connect": 1622209726934,
      "secureConnect": 1622209727148,
      "upload": 1622209727148,
      "response": 1622209727248,
      "end": 1622209727248,
      "phases": {
        "wait": 0,
        "dns": 1,
        "tcp": 72,
        "tls": 214,
        "request": 0,
        "firstByte": 100,
        "download": 0,
        "total": 387
      }
    },
    "message": "Response code 401 (UNAUTHORIZED)",
    "stack": "HTTPError: Response code 401 (UNAUTHORIZED)\n    at Request.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)",
    "options": {
      "headers": {
        "user-agent": "Renovate Bot (GitHub App 2740)",
        "authorization": "***********",
        "accept": "application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.docker.distribution.manifest.v2+json",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://quay.io/v2/<project>/manifests/latest",
      "username": "",
      "password": "",
      "method": "GET",
      "http2": false
    },
    "response": {
      "statusCode": 401,
      "statusMessage": "UNAUTHORIZED",
      "body": "{\"errors\":[{\"code\":\"UNAUTHORIZED\",\"detail\":{},\"message\":\"access to the requested resource is not authorized\"}]}\n",
      "headers": {
        "server": "nginx/1.12.1",
        "date": "Fri, 28 May 2021 13:48:47 GMT",
        "content-type": "application/json",
        "content-length": "112",
        "connection": "close",
        "docker-distribution-api-version": "registry/2.0",
        "www-authenticate": "Bearer realm=\"https://quay.io/v2/auth\",service=\"quay.io\",scope=\"repository:<project>:pull\""
      },
      "httpVersion": "1.1"
    }
  }
}
```

**Reply by giokara-work** [reply]
Looking into the `getAuthHeaders` function, a TODO is present "// TODO: debug why quay throws errors (#9604)". Possibly related?

**Reply by giokara-work** [reply]
However the PR is opened even with the error messages present. So this is already a great step forward, thanks for the support!

**Reply by HonkingGoose** [reply]
> @HonkingGoose Can you please add this sample to the docs if it is verified working.

@viceice Did we get things working properly yet? I'm not sure if I should start doing things or not yet... 😉

**Comment by lmilbaum** [comment]
@giokara-oqton I found this issue relevant to a problem I am hitting. Would you mind share if this solution is still working for you? I was unable to reconstruct it on my environment and I was hoping your information can help me solve my issue. 
Thanks in advace

**Reply by giokara-work** [reply]
The renovate.json configuration given by @viceice on May 27th above works for us (with error messages - PR is still opened though). If you use an encrypted token, as is the case in the solution, do not forget to actually encrypt the token with this tool: https://app.renovatebot.com/encrypt.

**Reply by viceice** [reply]
@lmilbaum i think we need some quay.io user who can provide some more relevant parts of a trace run.

Can you configure hostrules like https://github.com/renovatebot/renovate/discussions/10188#discussioncomment-792567 and try to update a single private image?
 Then provide the logs like we debugged your proxy issue.

**Reply by viceice** [reply]
- #9604
